src/GPGMailer.php

```diff
@@ -29,316 +29,153 @@
      * @param string $signKey           Key identifier, usually an email address but can be fingerprint
-     * @param string $signKeyPassphrase Optional passphrase for key required for signing
+     * @param string $signKeyPassphrase Passphrase required for the signKey
+     * @param array  $options           Option set {@see Crypt_GPGAbstract::__construct}
+     * @throws InvalidArgumentException
      */
-    public function __construct($encryptKey = null, $signKey = null, $signKeyPassphrase = null)
+    public function __construct($encryptKey = '', $signKey = '', $signKeyPassphrase = '', $options = [])
     {
-        parent::__construct();
-
-        // Set options
-        $this->setOptions();
-        $this->gpg = new Crypt_GPG($this->options);
-
-        // Add encryption key
-        if (is_null($encryptKey) && !defined('GPGMAILER_ENCRYPT_KEY')) {
-            throw new InvalidArgumentException('$encryptKey not defined');
-        }
-        $this->gpg->addEncryptKey($encryptKey ?: GPGMAILER_ENCRYPT_KEY);
-
-        // Add signing key
-        if ($signKey || defined('GPGMAILER_SIGN_KEY')) {
-            if (is_null($signKeyPassphrase) && defined('GPGMAILER_SIGN_KEY_PASSPHRASE')) {
-                $signKeyPassphrase = GPGMAILER_SIGN_KEY_PASSPHRASE;
-            }
-            $this->gpg->addSignKey($signKey ?: GPGMAILER_SIGN_KEY, $signKeyPassphrase);
-            $this->sign = true;
-        }
+        $this->setCryptGPGOptions($options);
+        $this->setEncryptKey($encryptKey ?: Environment::getEnv('GPGMAILER_ENCRYPT_KEY'));
+        $this->setSignKey($signKey ?: Environment::getEnv('GPGMAILER_SIGN_KEY'));
+        $this->setSignKeyPassphrase($signKeyPassphrase ?: Environment::getEnv('GPGMAILER_SIGN_KEY_PASSPHRASE'));
     }
 
     /**
-     * Set options for Crypt_GPG.
+     * Sets the encryption key for this mailer. If not set, the configured default will be used.
      *
-     * @see Crypt_GPGAbstract::__construct() for available options
+     * @param string|null $encryptionKey
+     * @return $this
      */
-    private function setOptions()
+    public function setEncryptKey($encryptionKey = null)
     {
-        $options = GPGMailer::config()->options;
-        if (isset($options[0]) && is_array($options[0])) {
-            $this->options = $options[0];
-        }
-
-        // Option to override home dir and provide a relative path instead
-        if (isset($this->options['relative_homedir'])) {
-            $this->options['homedir'] = Director::getAbsFile($this->options['relative_homedir']);
-            unset($this->options['relative_homedir']);
-        }
-
-        // Environment variables should override Configuration system
-        if (defined('GPGMAILER_HOMEDIR')) {
-            $this->options['homedir'] = GPGMAILER_HOMEDIR;
+        if (!isset($encryptionKey)) {
+            throw new InvalidArgumentException('Encryption key not defined');
```

Comment:
but in phpDoc above is written: "If not set, the configured default will be used "
and this setter is throwing exception and is allowing null ? confusing

Reply:
Removed confusing docblock :)
Automated setting makes no sense in this context.
Config can be used via dependency injection to call these setters with config values per environment :)